dbt/adapters/base/impl.py

```diff
@@ -1285,46 +1287,92 @@ def calculate_freshness(
+            else:
+                information_schema_to_metadata_sources[information_schema].append(source)
+
+        freshness_responses: Dict[BaseRelation, FreshnessResponse] = {}
+        for (
+            information_schema,
+            sources_for_information_schema,
+        ) in information_schema_to_metadata_sources.items():
+            result = self.execute_macro(
+                GET_RELATION_LAST_MODIFIED_MACRO_NAME,
+                kwargs={
+                    "information_schema": information_schema,
+                    "relations": sources_for_information_schema,
+                },
+                macro_resolver=macro_resolver,
+            )
+            adapter_response, table = result.response, result.table  # type: ignore[attr-defined]
+
+            for row in table:
+                try:
+                    last_modified_val = get_column_value_uncased("last_modified", row)
+                    snapshotted_at_val = get_column_value_uncased("snapshotted_at", row)
+                    identifier = get_column_value_uncased("identifier", row)
+                    schema = get_column_value_uncased("schema", row)
+                except Exception:
+                    raise MacroResultError(GET_RELATION_LAST_MODIFIED_MACRO_NAME, table)
+
+                freshness_response = self._create_freshness_response(
+                    last_modified_val, snapshotted_at_val
+                )
+                source_relation_for_result = schema_identifier_to_source[
+                    (schema.lower(), identifier.lower())
+                ]
+                freshness_responses[source_relation_for_result] = freshness_response
+
+        return adapter_response, freshness_responses
+
     def calculate_freshness_from_metadata(
         self,
         source: BaseRelation,
         macro_resolver: Optional[MacroResolverProtocol] = None,
     ) -> Tuple[Optional[AdapterResponse], FreshnessResponse]:
-        kwargs: Dict[str, Any] = {
-            "information_schema": source.information_schema_only(),
-            "relations": [source],
-        }
-        result = self.execute_macro(
-            GET_RELATION_LAST_MODIFIED_MACRO_NAME,
-            kwargs=kwargs,
+        adapter_response, freshness_responses = self.calculate_freshness_from_metadata_batch(
+            sources=[source],
             macro_resolver=macro_resolver,
         )
-        adapter_response, table = result.response, result.table  # type: ignore[attr-defined]
-
-        try:
-            row = table[0]
-            last_modified_val = get_column_value_uncased("last_modified", row)
-            snapshotted_at_val = get_column_value_uncased("snapshotted_at", row)
-        except Exception:
-            raise MacroResultError(GET_RELATION_LAST_MODIFIED_MACRO_NAME, table)
+        return adapter_response, list(freshness_responses.values())[0]
 
-        if last_modified_val is None:
+    def _create_freshness_response(
```

Comment:
I chose to carve out this helper out of `calculate_freshness_from_metadata_batch` as I can imagine adapter implementations that need to overwrite `calculate_freshness_from_metadata_batch` can still make use of this method for consistent FreshnessResponse creation. The other pieces of `calculate_freshness_from_metadata_batch` were pretty specific to the approach taken in the base